ethereum/_solidity.py

```diff
@@ -1,115 +1,375 @@
+# -*- coding: utf8 -*-
 import re
 import subprocess
 import os
-import yaml  # use yaml instead of json to get non unicode
+
+import yaml
+
+BINARY = 'solc'
 
 
 class CompileError(Exception):
     pass
 
 
-class solc_wrapper(object):
+def get_compiler_path():
+    """ Return the path to the solc compiler.
 
-    "wraps solc binary"
+    This funtion will search for the solc binary in the $PATH and return the
+    path of the first executable occurence.
+    """
+    for path in os.getenv('PATH', '').split(os.pathsep):
+        path = path.strip('"')
+        executable_path = os.path.join(path, BINARY)
 
-    @classmethod
-    def compiler_available(cls):
-        program = 'solc'
+        if os.path.isfile(executable_path) and os.access(executable_path, os.X_OK):
+            return executable_path
+
+    return None
+
+
+def get_solidity():
+    """ Return the singleton used to interact with the solc compiler. """
+    if get_compiler_path() is None:
+        return None  # the compiler wasn't found in $PATH
+
+    return solc_wrapper
+
+
+def solc_arguments(libraries=None, combined='bin,abi', optimize=True):
+    """ Build the arguments to call the solc binary. """
+    args = [
+        '--combined-json', combined,
+        '--add-std',
+    ]
+
+    if optimize:
+        args.append('--optmize')
+
+    if libraries is not None and len(libraries):
+        addresses = [
+            '{name}:{address}'.format(name=name, address=address)
+            for name, address in libraries.items()
+        ]
+        args.extend([
+            '--libraries',
+            ','.join(addresses),
+        ])
+
+    return args
+
+
+def solc_parse_output(compiler_output):
+    """ Parses the compiler output. """
+    result = yaml.safe_load(compiler_output)['contracts']
+
+    if 'bin' in result.values()[0]:
+        for value in result.values():
+            value['bin_hex'] = value['bin']
+
+            # decoding can fail if the compiled contract has unresolved symbols
+            try:
+                value['bin'] = value['bin_hex'].decode('hex')
+            except TypeError:
+                pass
+
+    for json_data in ('abi', 'devdoc', 'userdoc'):
+        # the values in the output can be configured through the
+        # --combined-json flag, check that it's present in the first value and
+        # assume all values are consistent
+        if json_data not in result.values()[0]:
+            continue
+
+        for value in result.values():
+            value[json_data] = yaml.safe_load(value[json_data])
+
+    return result
+
+
+def compiler_version():
+    """ Return the version of the installed solc. """
+    version_info = subprocess.check_output(['solc', '--version'])
+    match = re.search('^Version: ([0-9a-z.-]+)/', version_info, re.MULTILINE)
+
+    if match:
+        return match.group(1)
+
+
+def solidity_names(code):  # pylint: disable=too-many-branches
```

Comment:
This is function is used for backwards compatibility.

The `solc_wrapper.compile` function by default returns only the last contract, since the output of the compiler is unordered we need to inspect the source code. This version of the function was created to solve a couple of problems that had occurred:

- the contract and library keywords could not be at the start of the line
- the contract or library can be inside a comment or string

I'd like to deprecate the "return last contract" behavior, but for now I think it's required for backwards compatibility